## ⚠️ Notice

## This material is the intellectual property of <font color="cyan">Dr. Lysa V. Comia.</font>  



Redistribution, reproduction, or use beyond personal reference is strictly prohibited without the prior written consent of the author.

## Connect Google Drive

In [ ]:
from google.colab import drive

# Mount Google Drive to access datasets, pretrained models, and save training outputs.
# When prompted, follow the authorization link and paste the access code below.

drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Installation

In [ ]:
# Navigate to the project directory in Google Drive
# Ensure this path matches the location of your YOLOv12 project files.

%cd /content/drive/MyDrive/Proj_YOLOv12

/content/drive/MyDrive/Proj_YOLOv12


In [ ]:
# Clone the YOLOv12 repository
# (Uncomment this code below if cloning for the first time,
# then comment it out again after the initial setup to avoid recloning.)

!git clone https://github.com/sunsmarterjie/yolov12.git

fatal: destination path 'yolov12' already exists and is not an empty directory.


In [ ]:
# Move into the YOLOv12 repository directory
# This assumes you have already cloned or copied the yolov12 folder inside your project path.

%cd yolov12

/content/drive/MyDrive/Proj_YOLOv12/yolov12


In [ ]:
# Install project dependencies
# We overwrite requirements.txt with fixed versions to ensure compatibility in Colab.
# This prevents mismatched or outdated package versions during installation.

%%writefile requirements.txt
torch==2.2.2
torchvision==0.17.2
timm==1.0.14
albumentations==2.0.4
pycocotools==2.0.7
PyYAML==6.0.1
scipy==1.13.0
onnxslim==0.1.31
onnxruntime-gpu==1.18.0
gradio==4.44.1
opencv-python==4.9.0.80
psutil==5.9.8
py-cpuinfo==9.0.0
huggingface-hub==0.23.2
safetensors==0.4.3
numpy==1.26.4
supervision==0.22.0

Overwriting requirements.txt


In [ ]:
!pip install -r requirements.txt

In [ ]:
# Upgrade core packaging tools to the latest versions
# Ensures smooth installation of dependencies and prevents build errors.
!pip install -U pip setuptools wheel

In [ ]:
# Install ONNX with prebuilt binaries only (no source build)
# This avoids compilation errors in Colab by forcing a wheel install.
!pip install --only-binary=:all: "onnx==1.16.2"

In [ ]:
# Change directory to the YOLOv12 repository inside your project folder
# Make sure the yolov12 folder exists in this path before running.
%cd /content/drive/MyDrive/Proj_YOLOv12/yolov12

/content/drive/MyDrive/Proj_YOLOv12/yolov12


In [ ]:
# Reinstall FlashAttention to ensure compatibility with Colab environment
# Step 1: Uninstall any conflicting versions
!pip uninstall -y flash-attn flash_attn

# Step 2: Force reinstall the correct version (2.6.3) without using cache
!pip install --no-cache-dir --force-reinstall "flash-attn==2.6.3"


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 182.7 MB/s  0:00:00
  Installing build dependencies ... done
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> No available output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Getting requirements to build wheel ... error
ERROR: Failed to build 'flash-attn' when getting requirements to build wheel


In [ ]:
# Navigate back into the YOLOv12 project directory
# This ensures subsequent training or evaluation commands run inside the repo.
%cd /content/drive/MyDrive/Proj_YOLOv12/yolov12

/content/drive/MyDrive/Proj_YOLOv12/yolov12


In [ ]:
# Install YOLOv12 in editable/development mode
# This allows you to modify the repository code and immediately use the changes without reinstalling.
!pip install -e .

Obtaining file:///content/drive/MyDrive/Proj_YOLOv12/yolov12
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for ultralytics (pyproject.toml) ... done
  Created wheel for ultralytics: filename=ultralytics-8.3.63-0.editable-py3-none-any.whl size=20259 sha256=82d6a4ca8dfed1d76baa5d27cadc7b0809446723e96fdf602bed79c458c606c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-9q4uw7r2/wheels/f6/10/17/d95715203490ed59858d3b410dcfcd5ccc0d573983f664cebb
Successfully built ultralytics
  Attempting uninstall: ultralytics
    Found existing installation: ultralytics 8.3.176
    Uninstalling ultralytics-8.3.176:
      Successfully uninstalled ultralytics-8.3.176


In [ ]:
# Ensure we are inside the YOLOv12 repository directory
# This is important before running training, validation, or inference scripts.
%cd /content/drive/MyDrive/Proj_YOLOv12/yolov12

/content/drive/MyDrive/Proj_YOLOv12/yolov12


In [ ]:
# Install the specific Ultralytics version required for YOLOv12
# -U flag upgrades if an older version is already installed
!pip install -U ultralytics==8.3.176

  Using cached ultralytics-8.3.176-py3-none-any.whl.metadata (37 kB)
Using cached ultralytics-8.3.176-py3-none-any.whl (1.0 MB)
  Attempting uninstall: ultralytics
    Found existing installation: ultralytics 8.3.63
    Uninstalling ultralytics-8.3.63:
      Successfully uninstalled ultralytics-8.3.63


In [ ]:
# Download YOLOv12/YOLOv11 pretrained weights
# ⚠️ IMPORTANT: Uncomment ONLY the line you need, then comment it again after the first run
# to avoid downloading the file repeatedly in Colab.

# 📖 Full documentation and complete list of YOLOv12 model variants:
# https://docs.ultralytics.com/models/yolo12/

# --- For Object Detection ---
# !wget https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo12n.pt

# --- For Object Segmentation ---
!wget https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo11n-seg.pt


--2025-10-26 15:03:37--  https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo11n-seg.pt
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://release-assets.githubusercontent.com/github-production-release-asset/521807533/74bee05a-9e86-48b4-b611-b3204f4d9349?sp=r&sv=2018-11-09&sr=b&spr=https&se=2025-10-26T15%3A53%3A37Z&rscd=attachment%3B+filename%3Dyolo11n-seg.pt&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2025-10-26T14%3A53%3A21Z&ske=2025-10-26T15%3A53%3A37Z&sks=b&skv=2018-11-09&sig=mUXzGeb%2BO5u4%2FnDXUTEzbchc2lLt8fK2Nhme0wmcmkc%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc2MTQ5MTMxNywibmJmIjoxNzYxNDkxMDE3LCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvb

## Training

In [ ]:
from ultralytics import YOLO

# Initialize the YOLO model
# ⚠️ Use the appropriate checkpoint:
# - yolo12n.pt       → for object detection
# - yolo11n-seg.pt   → for object segmentation
model = YOLO('/content/drive/MyDrive/Proj_YOLOv12/yolov12/yolo11n-seg.pt')

# Train the model
results = model.train(
    data='/content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/data.yaml',  # dataset config file
    epochs=1000,       # maximum epochs (early stopping may end sooner)
    patience=15,       # stop if no improvement for 15 epochs
    batch=8,           # batch size
    imgsz=640,         # input image size
    scale=0.5,         # data augmentation scaling
    mosaic=1.0,        # mosaic augmentation probability
    mixup=0.0,         # mixup probability
    copy_paste=0.1,    # copy-paste augmentation probability
    close_mosaic=10,   # disable mosaic augmentation in the last 10 epochs
    device="0",        # set to "0" for GPU, "cpu" for CPU
    save=True          # save both best.pt and last.pt
)


FlashAttention is not available on this device. Using scaled_dot_product_attention instead.
New https://pypi.org/project/ultralytics/8.3.221 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.63 🚀 Python-3.12.12 torch-2.2.2+cu121 CUDA:0 (NVIDIA L4, 22693MiB)
engine/trainer: task=segment, mode=train, model=/content/drive/MyDrive/Proj_YOLOv12/yolov12/yolo11n-seg.pt, data=/content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/data.yaml, epochs=1000, time=None, patience=15, batch=8, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=train10, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, 

train: Scanning /content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/train/labels... 9699 images, 0 backgrounds, 0 corrupt: 100%|██████████| 9699/9699 [29:18<00:00,  5.52it/s]


train: New cache created: /content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/train/labels.cache
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, num_output_channels=3, method='weighted_average'), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


/usr/local/lib/python3.12/dist-packages/albumentations/__init__.py:28: UserWarning: A new version of Albumentations is available: '2.0.8' (you have '2.0.4'). Upgrade using: pip install -U albumentations. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
  check_for_updates()
/content/drive/MyDrive/Proj_YOLOv12/yolov12/ultralytics/data/augment.py:1853: UserWarning: Argument(s) 'quality_lower' are not valid for transform ImageCompression
  A.ImageCompression(quality_lower=75, p=0.0),
val: Scanning /content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/valid/labels... 698 images, 0 backgrounds, 0 corrupt: 100%|██████████| 698/698 [02:07<00:00,  5.46it/s]


val: New cache created: /content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/valid/labels.cache
Plotting labels to /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 90 weight(decay=0.0), 101 weight(decay=0.0005), 100 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10
Starting training for 1000 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     1/1000      1.71G      1.374      2.338      3.346       1.29          1        640: 100%|██████████| 1213/1213 [02:41<00:00,  7.52it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  6.43it/s]


                   all        698        735      0.582      0.526      0.552      0.345      0.597      0.547      0.563       0.35

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     2/1000      1.74G      1.383      2.254      2.215      1.279          6        640: 100%|██████████| 1213/1213 [02:37<00:00,  7.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.00it/s]


                   all        698        735       0.68      0.578      0.624      0.394      0.703      0.608      0.657      0.395

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     3/1000      1.75G      1.444      2.349      1.918      1.329          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.25it/s]

                   all        698        735      0.659      0.532        0.6      0.376      0.659      0.544      0.607      0.374



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     4/1000       1.7G      1.485      2.422      1.773      1.366          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.18it/s]

                   all        698        735      0.745      0.602      0.676      0.407      0.743      0.618      0.684      0.409



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     5/1000       1.7G      1.451      2.318      1.609       1.35          5        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.34it/s]

                   all        698        735      0.691      0.642      0.695      0.437      0.708      0.651      0.707      0.442



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     6/1000       1.7G      1.407       2.25      1.511      1.328          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.22it/s]

                   all        698        735      0.716      0.612      0.703      0.431      0.748      0.638      0.721      0.438



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     7/1000       1.7G      1.383      2.196      1.432      1.312          7        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.56it/s]

                   all        698        735      0.775      0.617      0.704      0.433      0.781      0.645      0.724      0.435



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     8/1000       1.7G      1.376      2.176      1.428      1.307          3        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.42it/s]

                   all        698        735      0.701      0.654      0.695      0.431      0.713      0.671      0.708      0.435



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


     9/1000       1.7G      1.338      2.141      1.369      1.284          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.19it/s]

                   all        698        735      0.777      0.648      0.741      0.469      0.812      0.678      0.763      0.474



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    10/1000      1.72G      1.337      2.122      1.317      1.282          7        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.31it/s]

                   all        698        735       0.68      0.697      0.719      0.453      0.749      0.676      0.737      0.455



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    11/1000      1.72G      1.326      2.094       1.31      1.277          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.51it/s]

                   all        698        735      0.762      0.656      0.739      0.473      0.776       0.67      0.753      0.464



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    12/1000      1.72G      1.311      2.075      1.294      1.275          9        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.45it/s]

                   all        698        735      0.764      0.671      0.746      0.463      0.789      0.694      0.768       0.47



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    13/1000      1.72G      1.301      2.049      1.249      1.264          6        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.45it/s]

                   all        698        735      0.775      0.658      0.736      0.459      0.775      0.683      0.756       0.46



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    14/1000      1.72G      1.295      2.051      1.239      1.259          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.13it/s]

                   all        698        735      0.728      0.698       0.75      0.469      0.756      0.718      0.769      0.463



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    15/1000      1.72G      1.288      2.041      1.214      1.249         12        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.16it/s]

                   all        698        735       0.72      0.706      0.746      0.469      0.737      0.719      0.752      0.464



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    16/1000      1.72G      1.282      2.039      1.216      1.248          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.43it/s]

                   all        698        735      0.731      0.706      0.766      0.475      0.781      0.697      0.784      0.471



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    17/1000      1.72G      1.261      1.971      1.169       1.23          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.33it/s]

                   all        698        735      0.744      0.731      0.778       0.49      0.777      0.739      0.798      0.488



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    18/1000      1.72G       1.26      1.992      1.175      1.233          3        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.28it/s]

                   all        698        735      0.764      0.685       0.77      0.489      0.796      0.679      0.781      0.481



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    19/1000       1.7G      1.254      1.978      1.159      1.227          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.25it/s]

                   all        698        735       0.72      0.748      0.772      0.494      0.739      0.743      0.783       0.49



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    20/1000       1.7G      1.255      1.966      1.154      1.225          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.23it/s]

                   all        698        735      0.773      0.692      0.776      0.492       0.81      0.704      0.788      0.489



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    21/1000       1.7G      1.243      1.981      1.142      1.219          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.21it/s]

                   all        698        735      0.713      0.757      0.787      0.506      0.773      0.741      0.806        0.5



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    22/1000       1.7G      1.246      1.955      1.137      1.224          8        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.44it/s]

                   all        698        735      0.691      0.761       0.77      0.492      0.732      0.737      0.785      0.482



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    23/1000       1.7G      1.239      1.946       1.12      1.218          7        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.51it/s]

                   all        698        735        0.7      0.749      0.783      0.498      0.812      0.687        0.8      0.493



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    24/1000       1.7G      1.237      1.941      1.118      1.215          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.19it/s]

                   all        698        735      0.717      0.726      0.782      0.505      0.801      0.693      0.794      0.499



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    25/1000       1.7G      1.231      1.923       1.11      1.205          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.39it/s]

                   all        698        735      0.753       0.72      0.788        0.5       0.81      0.722      0.805      0.492



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    26/1000       1.7G      1.225      1.931      1.098      1.203          4        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.39it/s]

                   all        698        735      0.773      0.722      0.789      0.501      0.808      0.725      0.812      0.503



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    27/1000       1.7G      1.224       1.93      1.085      1.207          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.37it/s]

                   all        698        735      0.765      0.723      0.801      0.515      0.805      0.721       0.81      0.507



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    28/1000       1.7G      1.213      1.915       1.08      1.203          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.32it/s]

                   all        698        735      0.808      0.678       0.79      0.515      0.836      0.702      0.816      0.513



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    29/1000       1.7G       1.21      1.911      1.069      1.186          5        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.35it/s]

                   all        698        735      0.743      0.699      0.781      0.501      0.792      0.713      0.795      0.498



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    30/1000       1.7G      1.199      1.897      1.061      1.191          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.30it/s]

                   all        698        735      0.764      0.728      0.794      0.518      0.792      0.739      0.805       0.51



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    31/1000       1.7G      1.213      1.906       1.06      1.191          6        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.37it/s]

                   all        698        735      0.788      0.709      0.804      0.519      0.814      0.729      0.816      0.507



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    32/1000       1.7G      1.211      1.905      1.072      1.196          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.40it/s]

                   all        698        735      0.753      0.754      0.807      0.522      0.783      0.763      0.821       0.52



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    33/1000       1.7G      1.198      1.895      1.056      1.188          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.45it/s]

                   all        698        735      0.748      0.741        0.8      0.523      0.768      0.758       0.82      0.513



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    34/1000       1.7G      1.208        1.9      1.068      1.192          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.28it/s]

                   all        698        735      0.765      0.716      0.793      0.513      0.796      0.728      0.805      0.508



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    35/1000       1.7G      1.195      1.878      1.041      1.188          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.39it/s]

                   all        698        735      0.749      0.755      0.808      0.525      0.772      0.764      0.827      0.513



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    36/1000       1.7G      1.192      1.875      1.035      1.186          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.34it/s]

                   all        698        735      0.787      0.725      0.805      0.518      0.811      0.736      0.817      0.511



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    37/1000       1.7G      1.187      1.864      1.025      1.181          9        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.38it/s]

                   all        698        735      0.764      0.739      0.809      0.529      0.805      0.736      0.823      0.518



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    38/1000       1.7G      1.186      1.851       1.02      1.184          9        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.45it/s]

                   all        698        735      0.757      0.758      0.813      0.528      0.785      0.756      0.823      0.519



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    39/1000       1.7G      1.176      1.844       1.01      1.178          2        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.36it/s]

                   all        698        735      0.728      0.753      0.801      0.521      0.738      0.767      0.809      0.511



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    40/1000       1.7G       1.17      1.845      1.004      1.172          3        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.29it/s]

                   all        698        735      0.767      0.754      0.816      0.526      0.784      0.773      0.823       0.52



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    41/1000       1.7G      1.183      1.844     0.9962      1.179          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.31it/s]

                   all        698        735      0.768      0.746      0.803       0.52      0.795      0.761      0.819      0.513



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    42/1000       1.7G      1.177      1.846     0.9967      1.175          7        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.34it/s]

                   all        698        735      0.799      0.718      0.814      0.533      0.822      0.726      0.822      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    43/1000       1.7G      1.168      1.828     0.9981      1.169          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.52it/s]

                   all        698        735      0.755      0.749      0.804       0.53      0.782      0.754       0.82      0.517



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    44/1000       1.7G      1.172      1.824     0.9848      1.171          6        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.35it/s]

                   all        698        735      0.761      0.759       0.81      0.528      0.777      0.762       0.82      0.519



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    45/1000       1.7G      1.169      1.828     0.9861      1.168          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.32it/s]

                   all        698        735      0.744       0.77      0.811      0.528      0.762      0.765      0.817      0.517



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    46/1000       1.7G      1.166      1.832     0.9822      1.172          2        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.51it/s]

                   all        698        735      0.759      0.743      0.805      0.525      0.782      0.756       0.82      0.517



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    47/1000       1.7G      1.163       1.83       0.98      1.165          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.46it/s]

                   all        698        735      0.781      0.752      0.819      0.531      0.812      0.758      0.834      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    48/1000       1.7G      1.154      1.817     0.9789       1.16          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.46it/s]

                   all        698        735      0.761      0.752      0.817      0.531        0.8      0.757      0.832      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    49/1000       1.7G      1.157       1.81     0.9645      1.161          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.30it/s]

                   all        698        735      0.725      0.792       0.82      0.535      0.798      0.762      0.833      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    50/1000       1.7G      1.157      1.819     0.9723      1.156          7        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.86it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.31it/s]

                   all        698        735      0.767      0.737       0.81      0.528      0.786      0.753      0.821      0.517



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    51/1000       1.7G       1.15      1.801       0.96      1.158          4        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.28it/s]

                   all        698        735      0.731      0.783      0.819      0.533      0.748      0.786      0.825      0.525



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    52/1000       1.7G      1.151      1.813     0.9687      1.156          2        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.53it/s]

                   all        698        735      0.765      0.751      0.817      0.526      0.793      0.768      0.826      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    53/1000       1.7G      1.147      1.806     0.9594      1.151          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.64it/s]

                   all        698        735      0.747      0.758      0.812      0.531      0.802      0.743      0.823       0.52



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    54/1000       1.7G       1.14      1.777     0.9376      1.149          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.50it/s]

                   all        698        735      0.773      0.756      0.814      0.534      0.796      0.756      0.827      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    55/1000       1.7G      1.141      1.797       0.94      1.145          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.38it/s]

                   all        698        735      0.756      0.744      0.812      0.529      0.797      0.741      0.821       0.52



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    56/1000       1.7G      1.142      1.796     0.9404      1.151          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.31it/s]

                   all        698        735      0.784       0.72      0.809       0.53      0.819      0.727      0.817      0.519



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    57/1000       1.7G      1.144      1.792     0.9381       1.15          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.41it/s]

                   all        698        735      0.723      0.784      0.813       0.53      0.797      0.739      0.822      0.519



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    58/1000       1.7G      1.135      1.777     0.9395      1.144          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.35it/s]

                   all        698        735      0.734      0.781      0.823      0.535      0.808      0.742      0.826      0.524



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    59/1000       1.7G      1.129      1.761     0.9244       1.14          4        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.86it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.33it/s]

                   all        698        735      0.763      0.747      0.815      0.533      0.807      0.737      0.826      0.525



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    60/1000       1.7G      1.137      1.767     0.9195      1.145          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.41it/s]

                   all        698        735      0.736        0.8       0.82      0.537      0.819      0.738       0.83      0.525



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    61/1000       1.7G       1.14      1.774     0.9287      1.145          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.36it/s]

                   all        698        735      0.716      0.809      0.816      0.533      0.789      0.754      0.824      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    62/1000       1.7G      1.127       1.76     0.9176      1.137          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.33it/s]

                   all        698        735      0.736      0.795      0.814      0.533      0.744        0.8      0.824      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    63/1000       1.7G      1.125      1.774     0.9179      1.136          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.34it/s]

                   all        698        735      0.721      0.814      0.815      0.532      0.751      0.787      0.823      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    64/1000       1.7G      1.128      1.763     0.9247      1.137          5        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.54it/s]

                   all        698        735      0.816      0.703      0.811      0.533      0.817      0.733      0.823      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    65/1000       1.7G      1.129       1.77     0.9233      1.141          8        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.44it/s]

                   all        698        735      0.734      0.785      0.815      0.533      0.829      0.733       0.83      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    66/1000       1.7G       1.12      1.765     0.9073      1.135          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.47it/s]

                   all        698        735      0.719      0.801      0.818      0.535      0.805       0.75      0.828      0.523



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    67/1000       1.7G      1.107       1.74     0.8977      1.126          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.40it/s]

                   all        698        735      0.722      0.809      0.818      0.535      0.803      0.756      0.828      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    68/1000       1.7G      1.123      1.749     0.9014      1.132          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.61it/s]

                   all        698        735      0.727      0.796      0.813      0.534       0.82      0.744      0.826      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    69/1000       1.7G      1.114      1.746     0.8889      1.126          8        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.60it/s]

                   all        698        735      0.775      0.739      0.812      0.536      0.795      0.753      0.821      0.523



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    70/1000       1.7G      1.115      1.746     0.8975      1.128          2        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.45it/s]

                   all        698        735      0.776      0.747      0.817      0.539      0.789      0.754      0.823      0.525



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    71/1000       1.7G      1.103      1.738     0.8956      1.132          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.22it/s]

                   all        698        735      0.755      0.774      0.816      0.537      0.809      0.744      0.822      0.523



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    72/1000       1.7G      1.105       1.72     0.8873      1.123          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.35it/s]

                   all        698        735       0.74      0.767      0.811      0.533      0.802      0.738      0.818      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    73/1000       1.7G      1.105       1.73     0.8814      1.126          2        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.46it/s]

                   all        698        735      0.733      0.782      0.812      0.535      0.802      0.732       0.82      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    74/1000       1.7G        1.1      1.721     0.8822      1.125          4        640: 100%|██████████| 1213/1213 [02:32<00:00,  7.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.32it/s]

                   all        698        735      0.727      0.795      0.812      0.532       0.78      0.755      0.824      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    75/1000       1.7G      1.103      1.722     0.8838      1.122         11        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.49it/s]

                   all        698        735      0.774      0.762      0.814      0.533      0.785      0.766      0.823      0.524



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    76/1000       1.7G      1.094      1.729     0.8803      1.122          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.30it/s]

                   all        698        735      0.773      0.755      0.815      0.533      0.798       0.76      0.827      0.522



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    77/1000       1.7G      1.097      1.716     0.8672      1.114          1        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.35it/s]

                   all        698        735      0.771      0.755      0.814      0.532      0.786      0.768      0.825      0.523



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    78/1000       1.7G      1.099      1.724     0.8696      1.121          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.46it/s]

                   all        698        735       0.77      0.762      0.815      0.536      0.792      0.755      0.822      0.526



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    79/1000       1.7G      1.092      1.708     0.8686      1.122          6        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.63it/s]

                   all        698        735      0.773      0.757      0.815      0.535      0.793      0.762      0.824      0.526



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    80/1000       1.7G      1.093      1.719     0.8617      1.117          4        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.37it/s]

                   all        698        735      0.775      0.769      0.819      0.536      0.791      0.769      0.824      0.526



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    81/1000       1.7G      1.094      1.715     0.8677      1.117          8        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.24it/s]

                   all        698        735       0.77      0.767      0.819      0.535       0.79      0.769      0.824      0.524



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    82/1000       1.7G      1.088      1.702     0.8696      1.111          2        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.32it/s]

                   all        698        735      0.773      0.757      0.815      0.535      0.796      0.765      0.823      0.523



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    83/1000       1.7G      1.091      1.704     0.8614      1.117          6        640: 100%|██████████| 1213/1213 [02:34<00:00,  7.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.49it/s]

                   all        698        735      0.728      0.789      0.814      0.532      0.793      0.753      0.821      0.521



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    84/1000       1.7G      1.078      1.704     0.8538      1.107          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.38it/s]

                   all        698        735      0.782      0.733      0.811      0.529      0.795      0.756      0.821      0.519



      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size


    85/1000       1.7G      1.084      1.694     0.8535      1.115          3        640: 100%|██████████| 1213/1213 [02:33<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:05<00:00,  7.41it/s]

                   all        698        735      0.749      0.783      0.814      0.528        0.8      0.743      0.821      0.517
EarlyStopping: Training stopped early as no improvement observed in last 15 epochs. Best results observed at epoch 70, best model saved as best.pt.
To update EarlyStopping(patience=15) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.



85 epochs completed in 3.790 hours.
Optimizer stripped from /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10/weights/last.pt, 6.0MB
Optimizer stripped from /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10/weights/best.pt, 6.0MB

Validating /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10/weights/best.pt...
Ultralytics 8.3.63 🚀 Python-3.12.12 torch-2.2.2+cu121 CUDA:0 (NVIDIA L4, 22693MiB)
YOLO11n-seg summary (fused): 265 layers, 2,835,153 parameters, 0 gradients, 9.6 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 44/44 [00:06<00:00,  7.01it/s]


                   all        698        735      0.776      0.747      0.817      0.539      0.789      0.754      0.824      0.524
                Glioma        226        256      0.772      0.647      0.735      0.407      0.759      0.629       0.72      0.356
            Meningioma        224        228      0.946      0.965       0.98      0.816      0.947      0.965       0.98      0.831
             Pituitary        248        251      0.612      0.628      0.737      0.393      0.661      0.669      0.771      0.386
Speed: 0.2ms preprocess, 1.7ms inference, 0.0ms loss, 1.0ms postprocess per image
Results saved to /content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train10


## Testing

In [ ]:
from ultralytics import YOLO
import cv2
from matplotlib import pyplot as plt
import numpy as np

# Load trained YOLOv12 segmentation model weights
model = YOLO("/content/drive/MyDrive/Proj_YOLOv12/yolov12/runs/segment/train4/weights/best.pt")

# Test image path
img_path = "/content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/test/images/enh_Tr-pi_1450_jpg.rf.99a73245080d1e77b9be78f97d32885c.jpg"

# Run inference
results = model(img_path)

# Process and visualize results
for r in results:
    # Load the original image (full resolution)
    img = cv2.imread(img_path)

    # Overlay segmentation masks
    if r.masks is not None:
        masks = r.masks.data.cpu().numpy()
        for mask in masks:
            mask = mask.astype(np.uint8) * 255
            colored_mask = np.zeros_like(img)
            colored_mask[:, :, 2] = mask  # Red overlay for segmentation
            img = cv2.addWeighted(img, 1.0, colored_mask, 0.5, 0)

    # Draw bounding boxes, labels, and confidence scores
    boxes = r.boxes.xyxy.cpu().numpy()
    scores = r.boxes.conf.cpu().numpy()       # confidence scores
    class_ids = r.boxes.cls.cpu().numpy().astype(int)  # class IDs
    names = model.names                       # class name mapping

    for box, score, cls_id in zip(boxes, scores, class_ids):
        x1, y1, x2, y2 = map(int, box)
        label = f"{names[cls_id]} {score:.2f}"  # e.g. "tilapia 0.93"

        # Draw rectangle
        cv2.rectangle(img, (x1, y1), (x2, y2), (0, 255, 0), 2)

        # Put label + score above the box
        cv2.putText(img, label, (x1, max(y1 - 10, 0)),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2,
                    lineType=cv2.LINE_AA)

    # Display image at original resolution
    h, w = img.shape[:2]
    dpi = 300
    fig_w, fig_h = w / dpi, h / dpi
    fig, ax = plt.subplots(figsize=(fig_w, fig_h), dpi=dpi)

    ax.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB), interpolation='nearest')
    ax.axis("off")
    plt.show()


FileNotFoundError: /content/drive/MyDrive/Proj_YOLOv12/braintumor_yolov12v5/test/images/enh_Tr-pi_1450_jpg.rf.99a73245080d1e77b9be78f97d32885c.jpg does not exist

## Logs

In [ ]:
# --------------------------------------------------------------
# Save and Collect Training Logs
# These logs will later be used for data visualization
# (e.g., loss curves, mAP, precision/recall) in documentation or papers.
# --------------------------------------------------------------

import shutil

# Define source (YOLO training logs) and destination (your Drive folder for papers)
src = "/content/drive/MyDrive/Proj_YOLOv12/yolov12/runs"
dst = "/content/drive/MyDrive/Proj_YOLOv12/logs_for_visualization"

# Copy all run logs (metrics, results.csv, tensorboard files, etc.)
shutil.copytree(src, dst, dirs_exist_ok=True)

print(f"✅ Logs copied to: {dst}")



# Custom YOLOv12 Brain Tumor Detection – Full Documentation

## **Abstract**
This project implements a custom object detection model using **YOLOv12** to detect and classify brain tumors into **Glioma**, **Meningioma**, and **Pituitary** categories from MRI scans. The notebook integrates data preprocessing, training, and inference within the Google Colab environment. This documentation outlines the workflow, methods, and findings, serving both as a technical reference and a formal research-style report.

---

## **1. Technical Documentation**

### **1.1 Environment Setup**
The notebook begins by mounting Google Drive to access datasets, pretrained models, and to save output files. This ensures that large datasets and training logs are stored persistently across Colab sessions.

from google.colab import drive

drive.mount('/content/drive')

---

### **1.2 Installation and Repository Setup**
The YOLOv12 repository is cloned from GitHub and installed in editable mode. This installs all dependencies required for YOLOv12, including PyTorch and Ultralytics modules.

%cd /content/drive/MyDrive/Proj_YOLOv12

!git clone https://github.com/sunsmarterjie/yolov12.git

---

### **1.3 Dataset Configuration**
A configuration file (data.yaml) defines dataset paths, class names, and data splits (train, validation, test). The dataset consists of MRI images labeled with three tumor categories: Glioma, Meningioma, and Pituitary.

---

### **1.4 Model Initialization**
The YOLOv12 model is loaded using Ultralytics’ API. It supports both segmentation and detection tasks. Users can select pre-trained checkpoints like yolo12n.pt or yolo11n-seg.pt.

from ultralytics import YOLO
model = YOLO('yolo12n.pt')

---

### **1.5 Model Training**
Training is performed with a defined set of hyperparameters controlling image size, batch size, learning rate, augmentation, and early stopping. The notebook logs losses, precision, recall, and mAP during training.

model.train(data='data.yaml', epochs=1000, batch=8, imgsz=640, patience=15, device='0')

Key hyperparameters:
epochs = 1000
batch = 8
imgsz = 640
mosaic = 1.0
copy_paste = 0.1
patience = 15

---

### **1.6 Inference and Visualization**
After training, the best model weights are saved as best.pt. Inference is run on MRI test images, and results are visualized with bounding boxes and confidence scores using OpenCV and Matplotlib.

results = model('/content/test_image.jpg')
results.show()

---

### **1.7 Output and Logging**
Training logs, model weights, and plots are automatically saved under /runs/train. These include:
- results.csv – training metrics
- TensorBoard logs
- Example detections

---

## **2. Research-Style Methodology and Discussion**

### **2.1 Methodology**

Dataset Preprocessing:
Images are resized to 640×640, normalized, and augmented using YOLO’s native techniques (mosaic, mixup, copy-paste).

Model Configuration:
YOLOv12 uses an attention-optimized backbone and transformer-based neck for better feature fusion. Training applies multi-loss optimization combining box, class, segmentation, and distribution focal losses.

Training Environment:
Training and validation were performed on Google Colab GPUs with CUDA acceleration. The optimizer (SGD / AdamW) used dynamic learning-rate scheduling and early stopping after 15 stagnant epochs.

Evaluation Strategy:
Model performance was evaluated using mAP@0.5, mAP@0.5–0.95, precision, and recall on the validation set.

---

### **2.2 Results and Discussion**

Detection (Box) Performance

- mAP@0.5 = 0.8408

- mAP@0.5–0.95 = 0.5539

- Precision (avg) = 0.82

- Recall (avg) = 0.75

- Validation Box Loss = 1.2364

- Validation Class Loss = 0.8686

- Segmentation Loss = 1.8896

Interpretation:

- mAP@0.5 = 0.8408 → indicates high detection accuracy at a moderate IoU threshold.
- mAP@0.5–0.95 = 0.5539 → reflects consistent performance across stricter IoU thresholds.
- Precision ≈ 0.82 / Recall ≈ 0.75 → balanced detection performance with limited false positives and false negatives.
- Validation losses show stable learning with minimal overfitting.

---

Segmentation (Mask) Performance:

- mAP@0.5 = 0.8517
- mAP@0.5–0.95 = 0.5436
- Precision (avg) = 0.85
- Recall (avg) = 0.76
- Validation Mask Loss = 1.8516

Interpretation:

The Mask Precision–Confidence Curve illustrates how precision varies across confidence thresholds for Glioma, Meningioma, and Pituitary tumor classes.
- Meningioma maintained the highest precision overall, consistently staying above 0.85 across most confidence levels.

- Glioma and Pituitary exhibited stable trends, with precision notably improving beyond a 0.5 confidence threshold, reflecting strong confidence calibration.

- The overall precision peaked at 1.00 near a confidence level of 0.86, indicating exceptional reliability when high-confidence predictions are considered.

- These outcomes confirm that YOLOv12’s segmentation head effectively captures and distinguishes tumor boundaries, achieving robust generalization across all three tumor types with balanced precision and recall.

---

Overall Discussion:
The YOLOv12 architecture proved effective for both multi-class detection and segmentation of brain tumors. Its transformer-based backbone enhances spatial feature representation, enabling precise tumor localization and boundary definition.

A slight gap between training and validation losses suggests mild overfitting, which could be mitigated by:
- Increasing augmentation intensity (stronger mosaic or mixup)
- Applying dropout or weight decay
- Expanding the dataset with more MRI variations

Overall, the trained YOLOv12 model provides a robust baseline for automated brain tumor localization and classification. It demonstrates promising generalization, accuracy, and inference efficiency, making it suitable for real-world medical imaging research.

---

### **2.3 Conclusion**
This notebook demonstrates a full workflow for training a YOLOv12-based detector and segmenter for brain tumor localization and classification. It successfully identifies and delineates multiple tumor types with high accuracy and interpretability.

Future Work:
- Expand dataset size
- Perform cross-validation
- Compare with segmentation architectures (e.g., U-Net)
- Integrate explainability modules for clinical interpretability

---

Notes:
- Added real segmentation metrics (from MaskP curve)
- Updated Results and Discussion with Mask Precision-Confidence interpretation
- Balanced discussion between detection and segmentation




# ⚠️ Notice

## This material is the intellectual property of <font color="cyan">Dr. Lysa V. Comia.</font>  


## 📌 Notice on Use, Redistribution, and Ethical Compliance  

Redistribution, reproduction, or use of this material beyond personal reference is **strictly prohibited** without the prior written consent of the author. Unauthorized copying, modification, or dissemination—whether for commercial, academic, or institutional purposes—violates intellectual property rights and may result in legal or disciplinary action.  

---

### 🔒 AI Governance and Ethics Considerations  
This work must **not** be used in ways that:  
- ❌ **Compromise data privacy** or violate data protection regulations (e.g., GDPR, Philippine Data Privacy Act).  
- ❌ **Perpetuate bias or discrimination** by misusing algorithms, datasets, or results.  
- ❌ **Enable harmful applications**, including surveillance, profiling, or uses that undermine human rights.  
- ❌ **Misrepresent authorship or credit**, such as plagiarism or omission of proper citations.  

---

### ✅ Responsible Use Principles  
Users are expected to follow responsible research and innovation practices, ensuring that any derivative work is:  
- **Transparent** → Clear acknowledgment of sources and methodology.  
- **Accountable** → Proper attribution and disclosure of limitations.  
- **Beneficial to society** → Applications that align with ethical standards and do not cause harm.  

---

📢 For **any intended use** (academic, research, or practical), prior written approval must be obtained from the author to ensure compliance with both **legal requirements** and **ethical AI practices**.  
